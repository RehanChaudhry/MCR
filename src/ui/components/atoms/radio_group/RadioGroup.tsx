import React, { useState } from "react";
import {
  Dimensions,
  StyleProp,
  StyleSheet,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
  ViewStyle
} from "react-native";
import { AppLabel } from "ui/components/atoms/app_label/AppLabel";
import usePreferredTheme from "hooks/theme/usePreferredTheme";
import { optimizedMemo } from "ui/components/templates/optimized_memo/optimized_memo";

export type Choice = { id: number; label: string };
type Props = {
  style?: StyleProp<ViewStyle>;
  values: Array<Choice>;
  onChange?: (value: Choice, index: number) => void;
  direction: DIRECTION_TYPE;
  byDefaultSelected?: number;
  itemsInRow?: number; //for horizontal buttons
  shouldNotOptimize?: boolean;
};

export enum DIRECTION_TYPE {
  HORIZONTAL = "horizontal",
  VERTICAL = "vertical"
}

export const RadioGroup = optimizedMemo<Props>(
  ({
    style,
    values,
    onChange,
    itemsInRow = 2,
    direction,
    byDefaultSelected = 0
  }) => {
    const theme = usePreferredTheme();
    const [selectedPosition, setSelectedPosition] = useState<number>(
      values.length - 1 >= byDefaultSelected ? byDefaultSelected : 0
    );

    function buttonPressed(position: number) {
      const oldSelectedOption = selectedPosition;
      setSelectedPosition(position);
      if (position !== oldSelectedOption) {
        onChange?.(values[position], position);
      }
    }

    function getStyleAsPerSelectionStatus(position: number) {
      return selectedPosition === position
        ? [{ backgroundColor: theme.themedColors.label }]
        : [{ backgroundColor: theme.themedColors.background }];
    }

    const getDirection = () => {
      if (direction === DIRECTION_TYPE.HORIZONTAL) {
        return styles.direction_horizontal;
      } else if (direction === DIRECTION_TYPE.VERTICAL) {
        return styles.direction_vertical;
      }
    };

    return (
      <View
        style={[style, getDirection(), styles.radioButtonWrapper]}
        testID={"RADIO_GROUP"}>
        {values.map((item, index) => (
          <View
            style={[
              styles.radioButtonContainer,
              {
                width: Dimensions.get("window").width / (itemsInRow - 10)
              }
            ]}
            key={item.label}>
            <TouchableOpacity
              testID={"RADIO_GROUP_BUTTON"}
              style={[
                styles.radioButton,
                {
                  borderColor: theme.themedColors.label,
                  backgroundColor: theme.themedColors.background
                }
              ]}
              onPress={() => buttonPressed(index)}>
              <View
                style={[
                  styles.radioButtonIcon,
                  getStyleAsPerSelectionStatus(index)
                ]}
              />
            </TouchableOpacity>
            <TouchableWithoutFeedback
              testID="RADIO_GROUP_LABEL"
              onPress={() => buttonPressed(index)}>
              <AppLabel style={styles.radioButtonText} text={item.label} />
            </TouchableWithoutFeedback>
          </View>
        ))}
      </View>
    );
  }
);

const styles = StyleSheet.create({
  radioButtonContainer: {
    flexDirection: "row",
    alignSelf: "flex-start",
    marginTop: 16
  },
  radioButtonWrapper: {
    flexWrap: "wrap",
    alignItems: "center",
    justifyContent: "flex-start",
    flex: 1
  },
  radioButton: {
    height: 20,
    width: 20,
    borderRadius: 10,
    borderWidth: 1,
    alignItems: "center",
    justifyContent: "center"
  },
  radioButtonIcon: {
    height: 16,
    width: 16,
    borderRadius: 10,
    justifyContent: "flex-start"
  },
  radioButtonText: {
    fontSize: 14,
    marginLeft: 10,
    marginRight: 16
  },
  direction_horizontal: {
    flexDirection: "row",
    flexWrap: "wrap",
    alignItems: "center"
  },
  direction_vertical: {
    flexDirection: "column"
  }
});
