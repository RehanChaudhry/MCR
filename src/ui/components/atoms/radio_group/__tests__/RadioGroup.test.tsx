import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import { DIRECTION_TYPE, RadioGroup } from "../RadioGroup";

test("check label match to the provided", () => {
  const { getByTestId } = render(
    <RadioGroup
      values={[
        { id: 1, value: "Male" },
        { id: 2, value: "Female" },
        { id: 3, value: "Others" }
      ]}
      direction={DIRECTION_TYPE.HORIZONTAL}
    />
  );
  const radioButton = getByTestId("RADIO_GROUP");
  expect(radioButton).toBeDefined();
});

test("check radio button on Press", () => {
  const onPress = jest.fn();
  const { getByText } = render(
    <RadioGroup
      values={[
        { id: 1, value: "Male" },
        { id: 2, value: "Female" },
        { id: 3, value: "Others" }
      ]}
      direction={DIRECTION_TYPE.HORIZONTAL}
      byDefaultSelected={1}
      onChange={onPress}
    />
  );
  const radioButton = getByText("Male");
  fireEvent.press(radioButton);
  expect(onPress).toBeCalledTimes(1);
  expect(onPress).toBeCalledWith({ id: 1, value: "Male" }, 0);
});

test("check radio button label on Press", () => {
  const onPress = jest.fn();
  const { getByText } = render(
    <RadioGroup
      values={[
        { id: 1, value: "Male" },
        { id: 2, value: "Female" },
        { id: 3, value: "Others" }
      ]}
      direction={DIRECTION_TYPE.HORIZONTAL}
      byDefaultSelected={2}
      onChange={onPress}
    />
  );
  const radioButton = getByText("Female");
  fireEvent.press(radioButton);
  expect(onPress).toBeCalledTimes(1);
  expect(onPress).toBeCalledWith({ id: 2, value: "Female" }, 1);
});

it("Snapshot testing", () => {
  const rendered = render(
    <RadioGroup
      values={[
        { id: 1, value: "Male" },
        { id: 2, value: "Female" },
        { id: 3, value: "Others" }
      ]}
      direction={DIRECTION_TYPE.HORIZONTAL}
    />
  ).toJSON();
  expect(rendered).toMatchSnapshot();
});
