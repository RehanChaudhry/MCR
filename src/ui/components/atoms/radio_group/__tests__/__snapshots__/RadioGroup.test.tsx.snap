// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexWrap": "wrap",
      },
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexWrap": "wrap",
        "justifyContent": "flex-start",
      },
    ]
  }
  testID="RADIO_GROUP"
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "flexDirection": "row",
          "marginStart": 10,
          "marginTop": 5,
          "paddingVertical": 10,
        },
        Object {
          "width": 250,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#000000",
          "borderRadius": 10,
          "borderWidth": 1,
          "height": 20,
          "justifyContent": "center",
          "opacity": 1,
          "width": 20,
        }
      }
      testID="RADIO_GROUP_BUTTON"
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 7,
              "height": 14,
              "justifyContent": "flex-start",
              "width": 14,
            },
            Array [
              Object {
                "backgroundColor": "#000000",
              },
            ],
          ]
        }
      />
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="RADIO_GROUP_LABEL"
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Array [
              Object {
                "fontFamily": "Inter-Regular",
              },
              Object {
                "color": "#000000",
              },
              Object {
                "fontSize": 16,
                "marginLeft": 10,
                "marginRight": 16,
              },
            ],
            Object {
              "fontFamily": "Inter-Regular",
            },
          ]
        }
      >
        Male
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "flexDirection": "row",
          "marginStart": 10,
          "marginTop": 5,
          "paddingVertical": 10,
        },
        Object {
          "width": 250,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#000000",
          "borderRadius": 10,
          "borderWidth": 1,
          "height": 20,
          "justifyContent": "center",
          "opacity": 1,
          "width": 20,
        }
      }
      testID="RADIO_GROUP_BUTTON"
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 7,
              "height": 14,
              "justifyContent": "flex-start",
              "width": 14,
            },
            Array [
              Object {
                "backgroundColor": "#FFFFFF",
              },
            ],
          ]
        }
      />
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="RADIO_GROUP_LABEL"
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Array [
              Object {
                "fontFamily": "Inter-Regular",
              },
              Object {
                "color": "#000000",
              },
              Object {
                "fontSize": 16,
                "marginLeft": 10,
                "marginRight": 16,
              },
            ],
            Object {
              "fontFamily": "Inter-Regular",
            },
          ]
        }
      >
        Female
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "flexDirection": "row",
          "marginStart": 10,
          "marginTop": 5,
          "paddingVertical": 10,
        },
        Object {
          "width": 250,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#000000",
          "borderRadius": 10,
          "borderWidth": 1,
          "height": 20,
          "justifyContent": "center",
          "opacity": 1,
          "width": 20,
        }
      }
      testID="RADIO_GROUP_BUTTON"
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 7,
              "height": 14,
              "justifyContent": "flex-start",
              "width": 14,
            },
            Array [
              Object {
                "backgroundColor": "#FFFFFF",
              },
            ],
          ]
        }
      />
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="RADIO_GROUP_LABEL"
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Array [
              Object {
                "fontFamily": "Inter-Regular",
              },
              Object {
                "color": "#000000",
              },
              Object {
                "fontSize": 16,
                "marginLeft": 10,
                "marginRight": 16,
              },
            ],
            Object {
              "fontFamily": "Inter-Regular",
            },
          ]
        }
      >
        Others
      </Text>
    </View>
  </View>
</View>
`;
